### PR TITLE
Handle UUIDs in the file metadata class

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #...
| Docs PR or issue | contao/docs#...

This PR adds a new `uuid` key to our file metadata, that gets set inside the `FigureBuilder` when it is defining meta data. 

As `Metadata` is immutable we now also have a `with()` method that allows merging by returning a new instance with the merged values.

### CODE REVIEW

**Review by Toflar — DISMISSED** [review]

File: core-bundle/src/File/Metadata.php [hunk 1]
```diff
@@ -68,6 +87,11 @@ public function getUrl(): string
         return $this->values[self::VALUE_URL] ?? '';
     }
 
+    public function getUuid(): ?string
```
Comment (ausi) [inline core-bundle/src/File/Metadata.php:93]: Should we add a doccomment that tells us if it is a binary 16-byte UUID or in the ASCII format?

File: core-bundle/src/Image/Studio/FigureBuilder.php [hunk 2]
```diff
@@ -597,14 +603,19 @@ private function onDefineMetadata(): ?Metadata
...
         // container from the default meta fields with empty values instead
         $metaFields = $this->filesModelAdapter()->getMetaFields();
 
-        return new Metadata(array_combine($metaFields, array_fill(0, \count($metaFields), '')));
+        $data = array_merge(
+            array_combine($metaFields, array_fill(0, \count($metaFields), '')),
+            $fileReferenceData
+        );
+
+        return new Metadata($data);
```
Comment (fritzmg) [inline core-bundle/src/Image/Studio/FigureBuilder.php:626]:
```suggestion
        $metadata = new Metadata(array_combine($metaFields, array_fill(0, \count($metaFields), '')));

        return $metadata->with($fileReferenceData);
```

May be? Then the merging is consistently handled via `Metadata::with`.

File: core-bundle/src/Image/Studio/FigureBuilder.php [hunk 3]
```diff
@@ -584,8 +584,14 @@ private function onDefineMetadata(): ?Metadata
...
 
+        $fileReferenceData = array_filter([
+            Metadata::VALUE_UUID => null !== $this->filesModel ? $this->filesModel->uuid : null,
+        ]);
+
         if (null !== $this->metadata) {
-            return $this->metadata;
+            return $this->metadata
+                ->with($fileReferenceData)
+            ;
```
Comment (fritzmg) [inline core-bundle/src/Image/Studio/FigureBuilder.php]:
```suggestion
            return $this->metadata->with($fileReferenceData);
```

It's a single line in line 606 as well, so may be a single line here as well?

File: core-bundle/src/Image/Studio/FigureBuilder.php (hunk 2 above)
Comment (m-vo) [reply core-bundle/src/Image/Studio/FigureBuilder.php:626]: Then you'd potentially create two instances vs one though. :fox_face:

File: core-bundle/src/Image/Studio/FigureBuilder.php (hunk 3 above)
Comment (m-vo) [reply core-bundle/src/Image/Studio/FigureBuilder.php]: The CS fixer did this. :cry:

File: core-bundle/src/File/Metadata.php (hunk 1 above)
Comment (m-vo) [reply core-bundle/src/File/Metadata.php:93]: Is the one inside `FilesModel->uuid` always in ASCII format or do we need to enforce this when setting it?

Comment (ausi) [reply core-bundle/src/File/Metadata.php:93]: I think `FilesModel->uuid` is always in binary format.

Comment (Toflar) [reply core-bundle/src/File/Metadata.php:93]: Let's enforce ASCII

File: core-bundle/src/Image/Studio/FigureBuilder.php (hunk 2 above)
Comment (fritzmg) [reply core-bundle/src/Image/Studio/FigureBuilder.php:626]: True, but that's the case with the previous returns anyway, isn't it?

File: core-bundle/src/File/Metadata.php (hunk 1 above)
Comment (ausi) [reply core-bundle/src/File/Metadata.php:93]: If we plan to use objects for UUIDs in the future we could return an uuid object here?
`\Symfony\Component\Uid\Uuid::fromString(...)`

File: core-bundle/src/Image/Studio/FigureBuilder.php (hunk 2 above)
Comment (m-vo) [reply core-bundle/src/Image/Studio/FigureBuilder.php:626]: The other two cases aren't build from scratch but from an explicitly set metadata object, though. There we need to go this route whereas in this case we're constructing the set on the fly.

I mean, at the end it's what you prefer, the memory + performance difference is negligible.

File: core-bundle/src/File/Metadata.php (hunk 1 above)
Comment (m-vo) [reply core-bundle/src/File/Metadata.php:93]: Done in 121f44875077ab1dce100ca2b1ed9245f5e9ff74 + e4a7bfee65eb89354e5614bc4807b2a6dc2c894b.

**Review by ausi — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]

**Review by Toflar — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**